### ISSUE TITLE
Integration wont load anymore on latest HA update 2023.6.3b

### ISSUE THREAD

**Issue by PV-Web** [bosch-thermostat/home-assistant-bosch-custom-component#301]
**Describe the bug**
Integration wont load anymore. in the notification i get this message on boot: 
The following integrations and platforms could not be set up:

[bosch](https://github.com/bosch-thermostat/home-assistant-bosch-custom-component) ([Show logs](https://10.0.0.158:8123/config/logs?filter=bosch))
Please check your config and [logs](https://10.0.0.158:8123/config/logs).

**To Reproduce**
Update to the latest version of HA 2023.6.b3

This is what the log says:
Logger: homeassistant.loader
Source: custom_components/bosch/__init__.py:12
Integration: bosch ([documentation](https://github.com/bosch-thermostat/home-assistant-bosch-custom-component), [issues](https://github.com/bosch-thermostat/home-assistant-bosch-custom-component/issues))
First occurred: 23:25:34 (1 occurrences)
Last logged: 23:25:34

Unexpected exception importing component custom_components.bosch
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/enum.py", line 260, in __set_name__
    enum_member._value_ = enum_class._member_type_(*args)
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/aioxmpp/xso/model.py", line 2923, in __init__
    self.xso_class = self._create_class()
                     ^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/aioxmpp/xso/model.py", line 2933, in _create_class
    self._create_name(),
    ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/aioxmpp/xso/model.py", line 2926, in _create_name
    return "".join(map(str.title, self.name.split("_")))
                                  ^^^^^^^^^
AttributeError: 'XSOEnumMixin' object has no attribute 'name'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/loader.py", line 813, in get_component
    ComponentProtocol, importlib.import_module(self.pkg_path)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1206, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1178, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1149, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/config/custom_components/bosch/__init__.py", line 12, in <module>
    from bosch_thermostat_client.const import (
  File "/usr/local/lib/python3.11/site-packages/bosch_thermostat_client/__init__.py", line 10, in <module>
    from .gateway import gateway_chooser
  File "/usr/local/lib/python3.11/site-packages/bosch_thermostat_client/gateway/__init__.py", line 1, in <module>
    from .ivt import IVTGateway, IVTMBLanGateway
  File "/usr/local/lib/python3.11/site-packages/bosch_thermostat_client/gateway/ivt.py", line 6, in <module>
    from bosch_thermostat_client.connectors import connector_ivt_chooser
  File "/usr/local/lib/python3.11/site-packages/bosch_thermostat_client/connectors/__init__.py", line 2, in <module>
    from .ivt import IVTXMPPConnector
  File "/usr/local/lib/python3.11/site-packages/bosch_thermostat_client/connectors/ivt.py", line 2, in <module>
    import aioxmpp
  File "/usr/local/lib/python3.11/site-packages/aioxmpp/__init__.py", line 101, in <module>
    from .errors import (  # NOQA
  File "/usr/local/lib/python3.11/site-packages/aioxmpp/errors.py", line 109, in <module>
    class ErrorCondition(structs.CompatibilityMixin, xso.XSOEnumMixin, enum.Enum):
  File "/usr/local/lib/python3.11/enum.py", line 557, in __new__
    raise exc
  File "/usr/local/lib/python3.11/enum.py", line 266, in __set_name__
    raise new_exc
** Debug SCAN **
** IMPORTANT **
Go to Developer tools in Home Assistant, choose Service tab and choose `bosch.debug_scan`
Download file to your computer and upload it somewhere eg. https://jsonblob.com/

**Comment by pszafer** [comment]
It's aioxmpp bug. We're waiting for upstream to fix https://github.com/horazont/aioxmpp/issues/391

**Comment by slovdahl** [comment]
Looks like it's actually a Python 3.11.3 regression: https://github.com/horazont/aioxmpp/issues/391#issuecomment-1575137853. Should be fixed in Python 3.11.4 that should be released tomorrow. Hopefully HA will update to it too.

**Comment by smoki3** [comment]
Looks like HA is already on 3.11.4. And not working for me

**Comment by slovdahl** [comment]
Are you sure? Python 3.11.4 has not even been released yet it seems.

https://www.python.org/downloads/

![image](https://github.com/bosch-thermostat/home-assistant-bosch-custom-component/assets/1417619/64141550-627f-4c73-8f8f-98080b62cff5)

**Comment by pszafer** [comment]
Python 3.11.4 hasn't been released yet, so impossible that your HA has it.

**Comment by smoki3** [comment]
oh sorry my fault, you are right

![image](https://github.com/bosch-thermostat/home-assistant-bosch-custom-component/assets/9723729/c1958ed0-3bb3-47be-ad5c-1f378c95a391)

**Comment by fisch55** [comment]
It seams that it will take longer to fix it….

**Comment by PV-Web** [comment]
Would there not be another way to temporarily fix this? until the new python is released?

**Comment by smoki3** [comment]
3.11.4 is released now. but not for HA looks like the will release 2023.6 with out the python update.

**Comment by fisch55** [comment]
> 3.11.4 is released now. but not for HA looks like they will release 2023.6 today without the python update.

Mad 😉

**Comment by slovdahl** [comment]
Let's hope we get a 2023.6.1 fairly soon with Python 3.11.4. 🤞

**Comment by JBakers** [comment]
Is there any other (temp) workaround maybe?
I got a couple of automations running from my Bosch entities, which all fail..

**Comment by gmed-ed** [comment]
I've been running in this issue.
Temporarly i returned to previous Version by using the backup made before updating.
After that, i hat to restart my KM200 device, because not all entities've been visible instandly.
Now everything is in working state again and time to wait for the phyton update.

**Comment by andyktaylor** [comment]
> I've been running in this issue. Temporarly i returned to previous Version by using the backup made before updating. After that, i hat to restart my KM200 device, because not all entities've been visible instandly. Now everything is in working state again and time to wait for the phyton update.

Yup in the same boat here, shame as the update has some pretty exciting changes, so hopefully not too long until we can move up.

**Comment by schlaufon** [comment]
I cannot confirm it's a python 3.11.x bug. 
see: _TypeError: _value_ not set in __new__, unable to create it
2023-06-09 07:21:34.691 ERROR (MainThread) [homeassistant.setup] Setup failed for custom integration bosch:_


Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/loader.py", line 813, in get_component
    ComponentProtocol, importlib.import_module(self.pkg_path)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1206, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1178, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1149, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/config/custom_components/bosch/__init__.py", line 12, in <module>
    from bosch_thermostat_client.const import (
  File "/usr/local/lib/python3.11/site-packages/bosch_thermostat_client/__init__.py", line 10, in <module>
    from .gateway import gateway_chooser
  File "/usr/local/lib/python3.11/site-packages/bosch_thermostat_client/gateway/__init__.py", line 1, in <module>
    from .ivt import IVTGateway, IVTMBLanGateway
  File "/usr/local/lib/python3.11/site-packages/bosch_thermostat_client/gateway/ivt.py", line 6, in <module>
    from bosch_thermostat_client.connectors import connector_ivt_chooser
  File "/usr/local/lib/python3.11/site-packages/bosch_thermostat_client/connectors/__init__.py", line 2, in <module>
    from .ivt import IVTXMPPConnector
  File "/usr/local/lib/python3.11/site-packages/bosch_thermostat_client/connectors/ivt.py", line 2, in <module>
    import aioxmpp
  File "/usr/local/lib/python3.11/site-packages/aioxmpp/__init__.py", line 101, in <module>
    from .errors import (  # NOQA
  File "/usr/local/lib/python3.11/site-packages/aioxmpp/errors.py", line 109, in <module>
    class ErrorCondition(structs.CompatibilityMixin, xso.XSOEnumMixin, enum.Enum):
  File "/usr/local/lib/python3.11/enum.py", line 557, in __new__
    raise exc
  File "/usr/local/lib/python3.11/enum.py", line 266, in __set_name__
    raise new_exc
TypeError: _value_ not set in __new__, unable to create it
2023-06-09 07:21:34.691 ERROR (MainThread) [homeassistant.setup] Setup failed for custom integration bosch: Unable to import component: Exception importing custom_components.bosch

**Comment by schlaufon** [comment]
It just won't load!
![image](https://github.com/bosch-thermostat/home-assistant-bosch-custom-component/assets/126504679/0522c9c7-a768-4453-834d-796cf40f5f53)

**Comment by smoki3** [comment]
https://github.com/home-assistant/docker-base/issues/221

**Comment by smoki3** [comment]
> It just won't load! ![image](https://user-images.githubusercontent.com/126504679/244600962-0522c9c7-a768-4453-834d-796cf40f5f53.png)

It's not loading because of this bug

**Comment by gmed-ed** [comment]
HA Core Update 2023.6.1 is showing up as update right now.
At a glance no detailled description about phyton minor version.

**Comment by Tuningmormota** [comment]
As temporary workaround the python can be updated to 3.11.4 manually ( i have done it on my HA docker image). After this integration works again.  But only do it if you know what you are doing, and how to go back if something goes south.

**Comment by frenck** [comment]
> So the update will be merged soon with 2023.6.2 it should be fixed

It depends on when it lands, I'm not sure if it will make into 2023.6.2 (or a patch release at all).

The linked PR to this issue is for bumping our base images, but we have more changes in the pipeline affecting those (cosign support).

Additionally, the linked PR is just the base image, which still needs to be built and shipped (which takes quite some hours). After that, I can start adjusting the Home Assistant base image, one that is tested, approved, build and shipped I can start upgrading the Home Assistant Core itself, which also needs to be tested, build and shipped.

TL;DR: Don't hold your breath on it being shipped in .2

../Frenck

**Comment by roderik-maker** [comment]
Thanks for keeping us updated and managing expectations! @frenck

**Comment by schlaufon** [comment]
> > So the update will be merged soon with 2023.6.2 it should be fixed
> 
> It depends on when it lands, I'm not sure if it will make into 2023.6.2 (or a patch release at all).

**@Frenck**

Thank you so much for caring about the issue.

May I ask, to include the recording sensors as well. Those have stopped working with a former update

Thank you soooo much in advance!

![image](https://github.com/bosch-thermostat/home-assistant-bosch-custom-component/assets/126504679/24d49c3e-0c42-452e-a839-1976e744824b)
.

**Comment by frenck** [comment]
@schlaufon I think you misunderstand the issue & solution, your request makes no sense to me. Sorry.

**Comment by schlaufon** [comment]
> @schlaufon I think you misunderstand the issue & solution, your request makes no sense to me. Sorry.

**@frenck**

Is that so. Then take my apology. 
Who -however- would you advice to contact for that recording sensor issue. 
Cheers!

**Comment by pszafer** [comment]
@schlaufon I will work on it when I get working docker image ;)

**Comment by JBakers** [comment]
> As temporary workaround the python can be updated to 3.11.4 manually ( i have done it on my HA docker image). After this integration works again. But only do it if you know what you are doing, and how to go back if something goes south.

Any suggestion as to how?

**Comment by gmed-ed** [comment]
you have to know, which parts ov pyghon are used by HA. Including dependencies etc.
Download those and put them in the right place within your docker-image.
Send a prayer to whatever gods you believe in and start the modified image.

Maybe it's easier to turn back to previous version via restoring backup or using an older docker image.

**Comment by Tuningmormota** [comment]
> @Tuningmormota
> 
> > As temporary workaround the python can be updated to 3.11.4 manually ( i have done it on my HA docker image). After this integration works again. But only do it if you know what you are doing, and how to go back if something goes south.
> 
> Any suggestion as to how? With the current hot weather here my temperature inside is not what it's supposed to be.

I have done the following:
1. Opened a consol to HA docker instance. (Run it Pi4, and portainer for docker managment)
2. I used the apk command the download what is needed to download, compile and bulid python (stackoverflow and google is recomended at this point, especiall for troubelshooting if needed)
3. donwload the  official python lib
4. extract it
5. Start worriing if it works at the end or not. (I have a backup every 30 minutes form my config folder, and weekly image of Pi SD card, so I was barve enough)
6. run  configure and make scripts and let it run for long time. (and hoping no errors will happen during the process)
7. restart docker image, and happly see that I see the temerature values for the rooms, and automation based on away mode are wiorking again.
8. check all other pages of HA is looks good or not. In my case yes.

Only do if you have a good quality backp, or you dont mind to start everthying from scratch. (as HA running, while under it you may exchange python libs which are currently in use. It very easily corrupts your HA database and config, and other thing which may different between our setups.) 
I did it bacause from 2023.6.1 I went back 2023.5.4 and I could reach energy and settings pages any more. Now I running  2023.6.1  with python 3.11.4 so far working. But I can update only to version which has python 3.11.4 already.

**Comment by gmed-ed** [comment]
one can see, it is possible.
Is it recomandable, i'm not sure
especially for a productive system.
Thanks for pointing the issue with upcoming updates need to have this python version, too

**Comment by JBakers** [comment]
@Tuningmormota 
Thanks!

**Comment by MajorTomD** [comment]
Hi there!

I am following this thread because my integration has stopped working too, of course, after the 2023.6 update. 
I am very new with Home Assistant and have absolutly no experience with all the details of coding, python and so on.

But I am not sure if I understand it correctly. The issue is due to the python update and there will be another update on HA, maybe not 2023.6.2, which will solve this problem?

I think this Bosch integration is a very cool one and I guess may most importend in HA. So I am very frustrated right now :)

Cheers
Sky

**Comment by slovdahl** [comment]
Correct, a Python update between HA 2023.5 and 2023.6 broke this integration. A newer Python version fixes it again, and we're hoping that that Python version will be available in HA 2023.6.2 or a later patch version, but there are no guarantees. The safest course of action in your case right now might be to go back to HA 2023.5 while we wait for the fixed Python version.

**Comment by mjmeierhoefer** [comment]
Is there any feedback that a newer python version will be used by homeassistant 2023.7?
Is there no easy way to fix the code to work with 2023.6?

**Comment by pszafer** [comment]
You can downgrade or try this https://github.com/bosch-thermostat/home-assistant-bosch-custom-component/issues/301#issuecomment-1586293676 or wait.

**Comment by rhehl** [comment]
a good description what to do for manuall upgrading python is the HA Docker file: https://github.com/home-assistant/docker-base/blob/master/python/3.11/Dockerfile

**Comment by rhehl** [comment]
@frenck if just restoring would work: yes

But sometimes it's not the best option.

If i try to rollback to 2023.05 i get the following error:

"Error during connection setup: [<class 'decimal.ConversionSyntax'>] (retrying in 15 seconds)
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/recorder/core.py", line 915, in _setup_recorder
    self._setup_connection()
  File "/usr/src/homeassistant/homeassistant/components/recorder/core.py", line 1394, in _setup_connection
    Base.metadata.create_all(self.engine)"

And i don't want to rollback the DB, this will make me loose my energy board history of the last 3 days.

**Comment by frenck** [comment]
> @frenck if just restoring would work: yes

If restoring didn't work, did you file an issue?

As in, making & restoring backups is extensively tested, and should just work. If not, we should definitely learn from that. It is important to work reliably.

> If i try to rollback to 2023.05 i get the following error:

Rolling back is not restoring a backup.

Your error is expected, as rolling back is not a supported/implemented thing in Home Assistant. We implement backward compatibility, not forward compatibility (like in this case, you are using a newer database format on an older HA version caused by the rollback).

**Comment by rhehl** [comment]
regarding newer databaser format: My schema_changes table said v41 changed 2023-04-07, v41 seems to be the latest and already 3 month old? https://github.com/home-assistant/core/blame/dev/homeassistant/components/recorder/db_schema.py (72).

due to the unchanged scheme, a rollback should have worked while retaining the db?

at least that would be my thought

In my special case, upgrading python to v3.11.4 solved my problems, with the slightest inconvenience

**Comment by gmed-ed** [comment]
In my case a restore back to 2023.5.6 was not a big issue.
After the restore completed i had to restart (not reset) my KM200 device and i was back in business again.

So backup/ restore is working as expected.

I got another HA Setup running at a different location without the Bosch integration so i could update to take a look for the new stuff.
For my needs i can go on for a while with 2023.05

**Comment by frenck** [comment]
> regarding newer databaser format: My schema_changes table said v41 changed 2023-04-07, v41 seems to be the latest and already 3 month old

@rhehl Rollbacks are not supported, and HA is not designed to work that way. It is not just the database format but also other data formats (e.g., config entries, entity registries, device registries, area registries, and even storage versions for individual integrations).

For that reason, please use backup to restore.

**Comment by slovdahl** [comment]
Good news if you didn't already notice: [2023.6.2 will have Python 3.11.4](https://github.com/home-assistant/core/pull/94556). Thank you very much @frenck! 🙇

**Comment by frenck** [comment]
Yup! @slovdahl! You are welcome.

I try to get a release out tomorrow (or at least before the weekend).

../Frenck

**Comment by JBakers** [comment]
Confirming core update 2023.6.2 fixed this issue!

**Comment by MOST2K2** [comment]
Yes its working again now! Thanks to all!

**Comment by Didel** [comment]
Looks like it is fixed with 2023.6.2. However, all 'devices' added by this integration are now duplicated in my installation. E.g. , the old `climate.woonkamer` entity does not work anymore, but a new `climate.woonkamer_2` entity is now also available that does work. Any ideas how to fix that? (using 0.22.2 of Bosch-termostat, and 2023.6.2 of HA)

**Comment by JBakers** [comment]
Delete the old ones and rename the new ones exactly the same?

**Comment by pszafer** [comment]
Issue is fixed now, closing it.

**Comment by 11Maarten** [comment]
@frenck 
Thank you very much. 
It works and my CT200 seems to report better then before.

**Comment by CharlieMK** [comment]
Updated, but heating is still showing as unavailable for me. Rebooted the system and reloaded the integration to no avail. Error:

```
Logger: homeassistant
Source: custom_components/bosch/__init__.py:388 
Integration: Bosch thermostat ([documentation](https://github.com/bosch-thermostat/home-assistant-bosch-custom-component), [issues](https://github.com/bosch-thermostat/home-assistant-bosch-custom-component/issues)) 
First occurred: 22:43:21 (1 occurrences) 
Last logged: 22:43:21

Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/config/custom_components/bosch/__init__.py", line 423, in thermostat_refresh
    await self.component_update(SWITCH, event_time)
  File "/config/custom_components/bosch/__init__.py", line 388, in component_update
    entities = self.hass.data[DOMAIN][self.uuid][component_type]
               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^
KeyError: 'switch'
```

**Comment by gmed-ed** [comment]
For mee, all went good.

Try to restart your heating device or at least the Bosch Gateway.
Maybe their Cloud didn't provide proper data.
Had a similar issue after restoring old HA Version.
